icu/icu4c/source/common/uprefs.cpp

```diff
@@ -0,0 +1,553 @@
+{
+    if (*status == U_UNSUPPORTED_ERROR)
+    {
+        warningGenerated = true;
+        *status = U_ZERO_ERROR;
+    }
+
+    if (uprv_strlen(secondData) != 0)
+    {
+        dest.append(firstData, *status);
+        dest.append(secondData, *status);
+    }
+}
+// -------------------------------------------------------
+// --------------- END OF HELPER FUNCTIONS ---------------
+// -------------------------------------------------------
+
+
+// -------------------------------------------------------
+// ---------------------- APIs ---------------------------
+// -------------------------------------------------------
+
+// Gets the valid and canonical BCP47 tag with the user settings for Language, Calendar, Sorting, Currency,
+// First day of week, Hour cycle, and Measurement system.
+// Calls all of the other APIs
+// Returns the needed buffer size for the BCP47 Tag. 
+int32_t uprefs_getBCP47Tag(char* uprefsBuffer, int32_t bufferSize, UErrorCode* status)
+{
+    RETURN_FAILURE_WITH_STATUS_IF(uprefsBuffer == nullptr && bufferSize != 0, U_ILLEGAL_ARGUMENT_ERROR, status);
+
+    *status = U_ZERO_ERROR;
+    CharString BCP47Tag;
+    bool warningGenerated = false;
+
+    CharString languageTag = getLocaleBCP47Tag_impl(status, false);
+    RETURN_VALUE_IF(U_FAILURE(*status), 0);
+    BCP47Tag.append(languageTag.data(), *status);
+    BCP47Tag.append("-u", *status);
+
+    CharString calendar = getCalendarSystem_impl(status);
+    RETURN_VALUE_IF(U_FAILURE(*status) && *status != U_UNSUPPORTED_ERROR, 0);
+    appendIfDataNotEmpty(BCP47Tag, "-ca-", calendar.data(), warningGenerated, status);
+    
+    CharString sortingSystem = getLocaleBCP47Tag_impl(status, true);
+    RETURN_VALUE_IF(U_FAILURE(*status) && *status != U_UNSUPPORTED_ERROR, 0);
+    appendIfDataNotEmpty(BCP47Tag, "-co-", sortingSystem.data(), warningGenerated, status);
+
+    CharString currency = getCurrencyCode_impl(status);
+    RETURN_VALUE_IF(U_FAILURE(*status) && *status != U_UNSUPPORTED_ERROR, 0);
+    appendIfDataNotEmpty(BCP47Tag, "-cu-", currency.data(), warningGenerated, status);
+
+    CharString firstDay = getFirstDayOfWeek_impl(status);
+    RETURN_VALUE_IF(U_FAILURE(*status) && *status != U_UNSUPPORTED_ERROR, 0);
+    appendIfDataNotEmpty(BCP47Tag, "-fw-", firstDay.data(), warningGenerated, status);
+
+    CharString hourCycle = getHourCycle_impl(status);
+    RETURN_VALUE_IF(U_FAILURE(*status) && *status != U_UNSUPPORTED_ERROR, 0);
+    appendIfDataNotEmpty(BCP47Tag, "-hc-", hourCycle.data(), warningGenerated, status);
+
+    CharString measureSystem = getMeasureSystem_impl(status);
+    RETURN_VALUE_IF(U_FAILURE(*status) && *status != U_UNSUPPORTED_ERROR, 0);
+    appendIfDataNotEmpty(BCP47Tag, "-ms-", measureSystem.data(), warningGenerated, status);
+
+    if (warningGenerated)
+    {
+        *status = U_USING_FALLBACK_WARNING;
+    }
+
+    int32_t result = checkBufferCapacityAndCopy(BCP47Tag.data(), uprefsBuffer, bufferSize, status);
```

Comment:
could this just be a direct return?

```suggestion
    return result = checkBufferCapacityAndCopy(BCP47Tag.data(), uprefsBuffer, bufferSize, status);
```